### ISSUE TITLE
Possible tf.matmul bug (wrong results) on tensorflow-gpu

### ISSUE THREAD

**Issue by sWizad** [tensorflow/tensorflow#31166]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v1.14.0-rc1-22-gaf24dc91b5 1.14.0
- Python version: 3.6.7
- GCC/Compiler version (if compiling from source): 8.2.0
- CUDA/cuDNN version: cuda-10.0
- GPU model and memory:

**Describe the current behavior**
tf.matmul on tensorflow-gpu gave wrong results. Here is a simplified version of the code.

```
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

r = [[ 1.0, 0.0],    [0.0, 1.0]]

x, y = np.meshgrid(list(range(400)), list(range(400)))
coords = np.stack([x,y],-1).reshape((400,400,2,1))
coords = tf.convert_to_tensor(coords,dtype=tf.float32)

r1 = tf.constant(r)

newCoords = tf.matmul(r1, coords)

sess = tf.Session()
ret = sess.run(newCoords,feed_dict={r1:r})

plt.matshow(ret[:,:,0,0])
plt.show()
```
When I ran it on my tensorflow-gpu, here is the result:
![bug](https://gist.githubusercontent.com/sWizad/3a25d6559ea3308e5cc2731519635c32/raw/2479341ed43fb8e69c94eda3a1362d5cbde7d2d7/Figure_1.png)
Looks like it stops computing halfway through and gave the rest 0 as a result.

**Describe the expected behavior**
Here is the result with CPU:
![CPU](https://user-images.githubusercontent.com/45821224/62137176-08631480-b2d5-11e9-8b8e-6348e8b206e7.png)

**********************************************************
Below is my old post. Initially, I thought the problem was related to TFRecord, but seems like this problem occurs without even using tfrecord too. For completeness, I keep the old example code with tfrecord.

```
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

def parser(serialized_example):
      fs = tf.io.parse_single_example(
          serialized_example,
          features={ "r": tf.FixedLenFeature([4], tf.float32) })
      fs["r"] = tf.reshape(fs["r"], [2, 2])
      return fs

r = [[ 1.0, 0.0],[0.0, 1.0]]

with tf.io.TFRecordWriter("cc.test") as tfrecord_writer:
    feature = {"r": tf.train.Feature(float_list=tf.train.FloatList(value=np.array(r).flatten() ))}
    example = tf.train.Example(features=tf.train.Features(feature=feature))
    tfrecord_writer.write(example.SerializeToString())
dataset = tf.data.TFRecordDataset(["cc.test"])
dataset = dataset.map(parser).repeat().make_one_shot_iterator()
features = dataset.get_next()

x, y = tf.meshgrid(list(range(400)), list(range(400)))
coords = tf.stack([x, y], -1)     #(h,w,2)
coords = tf.expand_dims(tf.cast(coords,tf.float32),-1) #(h,w,2,1)

r1 = features["r"]
r2 = tf.constant(r)

newCoords = tf.matmul(r1, coords)

sess = tf.Session()
ret = sess.run(newCoords[:,:,0,0])
plt.matshow(ret)
plt.show()
```
The code will create "cc.test" file to save the variable ``r`` and load it as ``r1``. Then matmul ``r1`` with some big varibles.

**Comment by ravikyram** [comment]
I have tried on colab with TF version 1.14 and was able to reproduce the issue.Please, find the [gist ](https://colab.research.google.com/drive/1-PILZtGYy4N5yV_goNWGx6CyOe22gutg)here.Thanks!

**Comment by jasonyzhang** [comment]
Replicated with Ubuntu 18.04, CUDA 10, tensorflow-gpu 1.14.0, and Python 2.7

**Comment by buaafish** [comment]
Replicated with 
Ubuntu 16.04, CUDA 10, tensorflow-gpu 1.14.0, and Python 2.7
Ubuntu 16.04, CUDA 10, tensorflow-gpu 1.14.0, and Python 3.6

**Comment by danmar3** [comment]
I'm having the same issue with:
Ubuntu 18.04, CUDA 10, tensorflow-gpu 1.14.0, Python 3.6, 
cudnn 7.6.2, driver version 430.26.

**Comment by BrianPugh** [comment]
I've also run into the same issue; current workaround is to perform the matmul on cpu.

**Comment by ravikyram** [comment]
@sWizad 

I tried in colab with TF GPU versions for 1.14, 1.15 and i am not seeing any issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/6c0d71133160fb0614bb28942c3fba04/untitled79.ipynb).Please, let us know whether we can close this issue.Thanks!

**Comment by humanpose1** [comment]
Hello,
I've run into the same issue on Ubuntu 18.04 CUDA 10.1 tensorflow-gpu version 2.0.0 and Python 3.7.4. Cudnn version is 7.6.2.
But in my case, if I use the version 1.14 of tensorflow the code above gives me the right answer.

It seems that in colab, there is no issue for the version 2.0 and for the 1.14.
Any idea why is there that problem ?

**Comment by fxtentacle** [comment]
@ravikyram @humanpose1  This is fixed on Colab because CUDA 10.1 is used there. However, the issue still persists in the official TF 1.15 or TF 2 python packages, which are using CUDA 10.0.

**Comment by fxtentacle** [comment]
This is most likely a duplicate of issue #26969.

**Comment by jvishnuvardhan** [comment]
@fxtentacle Can you please try with `TensorFlow 2.1.0-rc0` and let us know whether it resolved for you or not. Thanks!

**Comment by jvishnuvardhan** [comment]
I have tried `tf-nightly-gpu` and I cannot reproduce the issue. Please take a look at the [gist here](https://colab.sandbox.google.com/gist/jvishnuvardhan/d8c15cfca53160749ccf84c1443a3bbb/untitled679.ipynb). Thanks.
The issue is there in `TF1.15.0` and I am not sure whether there will be any more changes TF1.x. Thanks!

**Comment by fxtentacle** [comment]
@jvishnuvardhan Yes, `tf-nightly-gpu` works because it uses `CUDA 10.1` instead of `CUDA 10.0`. To confirm, I ran inside your gist:
```
from tensorflow.python.platform import build_info as tf_build_info
print(tf_build_info.cuda_version_number)
```
which returns `10.1`.

**Comment by fxtentacle** [comment]
BTW I believe the bug we're seeing here is this one:
https://docs.nvidia.com/cuda/archive/10.1/cuda-toolkit-release-notes/index.html#unique_1334956937
> In earlier releases, cuBLAS GEMM calls might randomly crash when running multiple host threads sharing one cuBLAS handle despite adhering to recommended usage in Thread Safety. This bug affects cuBLAS in earlier CUDA releases, and is fixed in CUDA 10.1. 

So this should be a very easy fix for `TF 1.15`, too: Just recompile with CUDA 10.1.

**Comment by jaingaurav** [comment]
@fxtentacle: Thank you for your analysis. As the fix is is part of CUDA 10.1, we recommend users to upgrade to a TF release with support for CUDA 10.1+. We do not plan to release a new 1.x based on CUDA 10.1.
